### DISCUSSION TITLE
Save the state of an fieldArray and then restore it via initial value.

### DISCUSSION THREAD

**Discussion by GTOsss** [react-hook-form/react-hook-form#1738] [General] [answered]
In submit or watch, I get the wrong object that can be passed to initialValues. How do I get the current state with an array of objects that contain the id?

At the moment my implementation looks like this. The problem is that I can't get the full state at the time of submit. 
```
const MyForm = () => {
  const [categories, setCategory] = React.useState<any>(formState.categories);
  const {register, handleSubmit, errors} = useForm({defaultValues: formState.categories});

  const addCategory = (name) => {
    setCategory(prev => [...prev, {name, key: uuid()}]);
  };

  const removeCategory = (id) => () => {
    setCategory(prev => [...prev.filter(item => item.key !== id)]);
  };

  const onSubmitInner = () => {
    onSubmit(1, {categories}, errors);
  };

  return (
    <div>
      <form onSubmit={handleSubmit(onSubmitInner)}>
        {categories.map((field, index) => (
          <div key={field.key} className={styles.fieldItem}>
            <Input
              className={styles.inputWrap}
              classNameInput={styles.input}
              register={register({required: true})}
              name={`categories[${index}]`}
              defaultValue={field.name}
              required
            />
            <Button onClick={removeCategory(field.id)} />
          </div>
        ))}
      </form>

      <AddFieldItem onAddItem={addCategory} placeholder="Add new category" />
    </div>
  );
});

**Comment by GTOsss** [comment]
I resolved my problem:
```
  const onSubmitInner = (values) => {
    const currentCategories = values.categories.map((value, i) => ({...categories[i], value}));
    onSubmit(1, {categories: currentCategories}, errors);
  };
```

But now I have another problem. Removing an element from the component state does not remove it from the state of the form itself.

**Comment by bluebill1049** [comment, marked as answer]
what do you mean by wrong object? can you supply a CSB?

**Reply by GTOsss** [reply]
In submit I get: 
``` {"categories":["test","b"]} ```

But for rendering I need objects with id:
 ```[{"value":"test","key":"f13e71a5"},{"value":"a","key":"c89c29d1"}]```
I had to store such an object separately in the state. This led to the need to implement deleting/changing/adding formItem yourself.

**Reply by GTOsss** [reply]
Perhaps hook useFieldArray would solve my problem by providing me with a valid state. But it doesn't work for me for some reason.

```
import * as React from 'react';
import {useForm, useFieldArray} from 'react-hook-form';

const Example = () => {
const {register} = useForm();
const {fields, append} = useFieldArray(
    {
      name: 'test',
    }
  );

  return (
    <div>
      {fields.map((field) => (
        <div key={field.id}>
          <input name="name" type="text" ref={register} />
        </div>
      ))}
      <button onClick={() => append({name: 'Test'})}>append</button>
    </div>
  );
};

export default Example;

```
```
react-hook-form.es.js:1498 Uncaught TypeError: Cannot read property 'control' of null
```

![image](https://user-images.githubusercontent.com/19191516/83348834-a500a380-a338-11ea-9fa5-55fff452cf16.png)

**Reply by GTOsss** [reply]
But as I see now in the documentation, it looks like useFieldArray won't solve my problem because it can't accept initialValue or initialState. It would be very cool if useFieldArray could give and receive a serializable reliable current state.

There are situations when you need to save exactly the state of the form (for example, in localStorage), but as I see, the form gives only formatted data suitable for creating a POST request, but it would be great to have a serialized state, which can later be passed to the form as intialState.

P.S. Perhaps if I could make useFieldArray work, I would solve this problem using intialValues

**Reply by bluebill1049** [reply]
Do you mean `defaultValues` at `useForm`?

**Reply by GTOsss** [reply]
Yes. But useFieldArray not working.

**Reply by bluebill1049** [reply]
codesandbox to reproduce?

**Reply by GTOsss** [reply]
https://codesandbox.io/s/react-hook-form-get-form-values-2llbl?file=/src/index.js

**Reply by GTOsss** [reply]
I realized what the mistake was. `control` must be passed if there is no context.
